Deprecate returning 408 for a server timeout on `_cluster/health`

arteam opened the pull request:
Add a deprecation warning and a system property `es.cluster_health.request_timeout_200` to opt in for returning 200 which will be the default in 8.0.0

Fixes #70849

arteam committed:
Return correct HTTP code for a server timeout on `_cluster/health`

Fixes #70849

arteam commented:
@elasticmachine update branch

arteam committed:
Check that we return 200 even in a case of a timeout

arteam committed:
Update Rest API tests

Make sure we don't get a 408 HTTP response code

arteam committed:
Update client tests for server cluster health timeouts

We should return 200 OK

arteam committed:
Don't catch the request timeout in a security test

The cluster health endpoint now returns 200 OK instead of 408

arteam committed:
Temporarily disable backward compatibility tests for 78180

arteam committed:
Return 408 REQUEST_TIMEOUT for ES7

arteam committed:
Mutes cluster health tests

arteam committed:
cluster-health tests don't belong in x-pack

arteam committed:
Fix the name of the ignored compatibility tests

arteam committed:
Disable BWC for #78180

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam committed:
Revert "Disable BWC for #78180"

This reverts commit 14d92a93ef96a704f286fbfa7fb96347143894a1.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam committed:
Skip bwc tests for the request timeout

arteam committed:
Add deprecation warning for a cluster health timeout

arteam committed:
Update deprecation message

arteam committed:
Return REQUEST_TIMEOUT for v7

arteam committed:
Add a system property to opt out from returning 200 on cluster health timeout

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam committed:
Merge remote-tracking branch 'origin/master' into return-200-for-cluster-health-timeout

arteam committed:
Ignore cluster health tests for rest compatibility

arteam committed:
Skip security tests for cluster timeout

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -299,7 +304,14 @@ public String toString() {
+                "HTTP status code for an internal cluster health timeout will change from 408 to 200 in a future major version");
+        }
+        if (Boolean.parseBoolean(System.getProperty("es.cluster_health.request_timeout_408", "false"))) {
+            return RestStatus.REQUEST_TIMEOUT;
```
This returns `REQUEST_TIMEOUT` even if we didn't time out?

arteam replied:
Yes, indeed, the check wasn't put correctly. I need to add a test that sets `es.cluster_health.request_timeout_408` to catch such mistakes.

DaveCTurner replied:
++ I'd suggest adding another test to the `qa` project, that's generally the simplest way to control things like system properties.

arteam replied:
I've decided to make the property as not static, so we can set in a constructor in `ClusterHealthResponsesTests`. It seems to much faster and easier to test this way instead of creating another full blown integration test.

DaveCTurner replied:
The trouble with this is that if the system property is invalid, it'll just make cluster health explode (with a `500 Internal Server Error`) on a timeout. We should be stricter and not even let the node start up in that case.

I'm ok with passing the flag into the response constructor for a unit test but I think we still have to parse the property up front while the node is initialising, and I would very much prefer to have an end-to-end test that the opt-in actually works.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -299,7 +304,14 @@ public String toString() {
-        return isTimedOut() ? RestStatus.REQUEST_TIMEOUT : RestStatus.OK;
+        if (RestApiVersion.current() == RestApiVersion.V_7 && isTimedOut()) {
+            deprecationLogger.critical(DeprecationCategory.API, "cluster_health_request_timeout",
+                "HTTP status code for an internal cluster health timeout will change from 408 to 200 in a future major version");
```
It's best for messages like this to be actionable: say that users should set the `es.cluster_health.request_timeout_408` system property to opt in to the new behaviour.

Also if they _have_ opted in to the new behaviour we should not emit a warning.

arteam replied:
I think I got confused whether the user should opt-in or opt out from the new behaviour. Do we need two properties: one for 7.x that allows user to opt-in for the 200 OK code and one in 8.x to opt out from it and keep returning `408`?

DaveCTurner replied:
No, just the opt-in property for 7.x. In 8.x you opt out via the REST compatibility layer. However it works best to make breaking changes like this with two PRs: the first implements the deprecation and gets merged to **both** `7.x` and `master`, and then the second reverts the deprecation and implements the actual change in `master` alone. Merging the deprecation to both branches is non-obvious but it has the advantage that if the 8.0 change slips then it doesn't introduce a regression, we just keep it deprecated for longer.

arteam replied:
Right, I didn't realize that we need to check the version from `RestRequest.getRestApiVersion`, not `RestApiVersion.current()`. This way the user can enable the v7 compatibility mode.

DaveCTurner replied:
:+1: tbc I don't think we need to check the `RestApiVersion` at all in the first PR (the one that does the deprecation), only in the second one.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -299,7 +304,14 @@ public String toString() {
     @Override
     public RestStatus status() {
-        return isTimedOut() ? RestStatus.REQUEST_TIMEOUT : RestStatus.OK;
+        if (RestApiVersion.current() == RestApiVersion.V_7 && isTimedOut()) {
```
This is always `false` isn't it? (In `master` at least)

arteam replied:
Yes. I'm not sure what the correct way to handle this. Should I add an explicit check or create 2 different PRs for the master and 7.x branches?

DaveCTurner replied:
See previous comment: deprecate this (and introduce the opt-in property) in both branches and then switch to the new behaviour in `master`.

arteam replied:
Done!

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -299,7 +304,14 @@ public String toString() {
+            deprecationLogger.critical(DeprecationCategory.API, "cluster_health_request_timeout",
+                "HTTP status code for an internal cluster health timeout will change from 408 to 200 in a future major version");
+        }
+        if (Boolean.parseBoolean(System.getProperty("es.cluster_health.request_timeout_408", "false"))) {
```
Best not to parse the property each time, instead we should have a static constructor which reads it into a static field. Moreover we should reject any value except `true` in that static constructor.

arteam replied:
Good point! I've moved the property to a static final field. Is there any standard ES way to validate the property to make sure we reject any value except `true`?

DaveCTurner replied:
Yes, but I think you need a static constructor rather than doing it in a one-liner. See e.g. this:

https://github.com/elastic/elasticsearch/blob/4ef3a58a46cc51abe0fb213d2502f21ad143fc58/server/src/main/java/org/elasticsearch/transport/TransportService.java#L74-L84

Also an invalid value should fail the node, but I don't think we construct the `ClusterHealthResponse` class early enough in the node's lifecycle to do that. `TransportClusterHealthAction` is likely a better bet.

arteam replied:
Done! I've decided to extract to a method.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -299,7 +304,14 @@ public String toString() {
+        if (Boolean.parseBoolean(System.getProperty("es.cluster_health.request_timeout_408", "false"))) {
+            return RestStatus.REQUEST_TIMEOUT;
+        }
+        return RestStatus.OK;
```
If the user requested v7 compatibility and we timed out then we should keep the old behaviour. That would mean there's no need to skip any tests in v7-compatible mode.

arteam replied:
Is there a way to set a system property for the REST compatibility tests?

DaveCTurner reviewed (changes requested):
I'm confused by the logic, maybe I'm missing something. I left a handful of comments.

arteam committed:
Make the ClusterHealthResponse change backward-compatible

For the version 7, we should always return 408 and a new parameter
to opt-in for 200.

arteam committed:
The parameter should provide a way for a user to opt OUT from the new behaviour

arteam committed:
Remove rest api version check

arteam committed:
Rollback 408 to 200 change

arteam committed:
Fix the deprecation message to use the opt in patameter

arteam committed:
Add tests for deprecation messages

arteam committed:
Revert "Update client tests for server cluster health timeouts"

This reverts commit 3a572d448583ba5628ae3d7674e6f74ac851dc4b.

arteam committed:
Fix checkstyle warning

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam committed:
Accept only true for esClusterHealthRequestTimeout200

arteam committed:
Satisfy checkstyle

arteam committed:
Update the variable holding the deprecation message

arteam committed:
Add warning assertion

arteam committed:
Add a test for opting in for the 200 status code

arteam committed:
Rename to readEsClusterHealthRequestTimeout200FromProperty

DaveCTurner commented on client/rest-high-level/src/test/java/org/elasticsearch/client/ClusterClientIT.java:
```diff
@@ -303,6 +303,8 @@ public void testClusterHealthNotFoundIndex() throws IOException {
         assertThat(response.status(), equalTo(RestStatus.REQUEST_TIMEOUT));
         assertThat(response.getStatus(), equalTo(ClusterHealthStatus.RED));
         assertNoIndices(response);
+        assertWarnings("HTTP status code for an internal cluster health timeout will be changed from 408 to 200 in 8.0.0. " +
+            "Set the `es.cluster_health.request_timeout_200` property to `true` to opt in for it.");
```
Why "internal" here? Also let's just say "a future version" in case the breaking change doesn't land by 8.0.

```suggestion
        assertWarnings("The HTTP status code for a cluster health timeout will be changed from 408 to 200 in a future version. " +
            "Set the `es.cluster_health.request_timeout_200` system property to `true` to suppress this message and opt in to the future behaviour now.");
```

(probably needs reformatting for line length, sorry there's no easy way to check for that in Github)

DaveCTurner reviewed (commented):
I suggested a slight rewording of the deprecation warning + continued the thread about handling the system property

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -97,7 +100,10 @@
-            (XContentParser parser, Void context, String index) -> ClusterIndexHealth.innerFromXContent(parser, index);
+        (XContentParser parser, Void context, String index) -> ClusterIndexHealth.innerFromXContent(parser, index);
+    static final String CLUSTER_HEALTH_REQUEST_TIMEOUT_DEPRECATION_MSG = "HTTP status code for an internal cluster health timeout " +
+        "will be changed from 408 to 200 in 8.0.0. Set the `es.cluster_health.request_timeout_200` property to `true` to opt in for it.";
```
Oh also it's better to extract the property name to a field somewhere and then use the field in messages instead of using the literal string in multiple places, see e.g. `PERMIT_HANDSHAKES_FROM_INCOMPATIBLE_BUILDS_KEY`. Saves worrying about typos and helps make sure we don't accidentally have any messages that refer to a nonexistent setting.

And one more thing: we prefer wrapping literals in `[]` rather than backticks in messages like this, backticks tend to result in copy/paste errors whereas brackets are harder to miss:

```suggestion
        "will be changed from 408 to 200 in 8.0.0. Set the [" + PROPERTY + "] property to [true] to opt in for it.";
```

arteam committed:
Update client/rest-high-level/src/test/java/org/elasticsearch/client/ClusterClientIT.java

Co-authored-by: David Turner <david.turner@elastic.co>

arteam committed:
Extract the property name to a constant

arteam committed:
Reformat error message

arteam commented:
@elasticmachine run elasticsearch-ci/rest-compatibility

arteam committed:
Adapt the warning message with square brackets

arteam committed:
Add the key prefix to the constant

arteam committed:
Revert "Adapt the warning message with square brackets"

This reverts commit fcee4c577d9a615cec115f1c77c3d5ed5cf087ff.

arteam committed:
Use brackets

arteam committed:
Update the error message in ClusterHealthResponse

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

arteam committed:
Add DeprecationLogger import

DaveCTurner reviewed (approved):
One nit, otherwise LGTM

arteam committed:
Update server/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java

Co-authored-by: David Turner <david.turner@elastic.co>

arteam committed:
Update the error message for ClusterClientIT

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-200-for-cluster-health-timeout

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
There are no branches to backport to. Aborting.

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 78180`
